File: java/recursion/TowerOfHanoi.java
package recursion;

public class TowerOfHanoi {

    private void performAction(int numberOfRings, String towerA, String towerB, String towerC) {
        if (numberOfRings > 0) {
            performAction(numberOfRings - 1, towerA, towerC, towerB);

            System.out.println("Moving Ring from " + towerA + " to " + towerC);

            performAction(numberOfRings - 1, towerB, towerA, towerC);
        }
    }
    public static void main(String[] args) {
        new TowerOfHanoi().performAction(4, "A", "B", "C");
    }
}

File: cpp/recursion/towerofhanoi.cpp
#include<iostream>

using namespace std;

void towerOfHanoi(int numberOfRings, int a, int b, int c) {
    if(numberOfRings > 0) {
        towerOfHanoi(numberOfRings - 1, a, c, b);
        cout << "moving ring from " << a << " to " << c << endl;
        towerOfHanoi(numberOfRings - 1, b, a, c);
    }
}

int main(int argc, char const *argv[])
{
    towerOfHanoi(5, 1, 2, 3);
    return 0;
}
